### DISCUSSION TITLE
Resolving interdependent constraints among fields

### DISCUSSION THREAD

**Discussion by seh** [cuelang/cue#971] [General]
Following up from [previous discussion in the "language" channel](https://cuelang.slack.com/archives/CLT4FD7KP/p1620940382109100) of the "CUE" Slack workspace, I'm interested in probing how CUE can resolve certain interdependencies among attributes—what I think of as _propagator network_ examples—but not others. For example, it can handle arithmetic:
```cue
a: b - 1
b: a + 1
c: a + b
```
You can supply any two of the _a_, _b_, and _c_ fields and CUE can figure out the missing third.

With a somewhat analogous example using strings, consider the following:
```cue
import "strings"
#thing: {
  a: string
  b: string
  result: strings.Join([a, b], ".")
  let parts = strings.Split(result, ".")
  a: parts[0]
  b: parts[1]
}
x: #thing & {
  a: "foo"
//  b: "bar"
  result: "foo.bar"
}
```
If you leave out field _a_ or _b_ and include _result_, CUE can figure out the missing one. If you leave out _result_, though, CUE doesn’t come up with a complete value for it. I can see some circularity in how _result_ depends partly on _a_, and _a_ depends on part of _result_. Why doesn't that problem afflict the first example as well?

**Comment by myitcv** [comment]
Copying across @mpvl's response from Slack

> To solve the "second" case, you would would, in the general case, have to implement an algebraic solver. In this case the solution is simple, but in the general case, this is not simple or even solvable. Adding support for it would make it very hard to explain where the boundaries lie on what is solvable and what not. Moreover, it would likely become a moving target and source for cross-version incompatibility.
> The easiest line to draw is to simply not do any algebraic solving. So CUE makes no attempt at algebraic solving whatsoever. Neither do typical propagator networks, IIUC: you would have to manually specify all directions.
> Of course one could always where such a solver as a third-party tool, which then injects the rewritten formulas in the configuration, possibly in cue.mod/usr.

The punchline here being that CUE will not add support for algebraic solving, but that a third-party tool could do such a thing with CUE as a basis for the approach.

I would be _hugely_ interested in such a tool/exploration if someone were to start it.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/971.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
